### PR TITLE
WIP,MNT: Add support for QtPy

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
This PR adds support for `qtpy` which allows us to become Qt-agnostic. This can eventually make the transition to Qt6 easier.

Related to https://github.com/mne-tools/mne-python/issues/9710#issuecomment-1063371893

### CODE REVIEW

**Comment by GuillaumeFavelier** [comment]
As I explain in https://github.com/mne-tools/mne-python/issues/9710#issuecomment-1063942500, the resource system needs some rework to allow Qt6 support.

**Comment by GuillaumeFavelier** [comment]
Although there is no `pyrcc6`, `pyside6-rcc` does exist:

https://doc.qt.io/qtforpython/tutorials/basictutorial/qrcfiles.html

So I think we could migrate to `PySide6` easily but for the sake of being Qt-agnostic, I still plan to rework the resource system.

**Comment by larsoner** [comment]
Could we just load the PNGs directly?

**Comment by GuillaumeFavelier** [comment]
@larsoner feel free to take a look at the code. This did the trick to me for PySide6, I'll try with PyQt6 next.

> Could we just load the PNGs/SVGs directly?

Definitely, but I think we need to be smart here because a relative path won't do. I chose `importlib.resources` and found that it worked quite well. I just refactored `_init_qt_resources()`.

Of course, the PR is not ready, I tried `Brain` and `coreg` mainly.

**Comment by GuillaumeFavelier** [comment]
There are still many places in the project where `PyQt5` is used or referred to directly:

<details>

```
.circleci/config.yml:            name: Check PyQt5
.circleci/config.yml:            command: LD_DEBUG=libs python -c "from PyQt5.QtWidgets import QApplication, QWidget; app = QApplication([])"
doc/install/mne_python.rst:    PyQt5:         5.14.1
mne/conftest.py:    # Check PyQt5
mne/conftest.py:        import PyQt5  # noqa: F401
mne/conftest.py:        pytest.skip('PyQt5 is not installed but needed for pyqtgraph!')
mne/conftest.py:        pytest.skip(f'PyQt5 has version {_check_pyqt5_version()}'
mne/conftest.py:        pytest.skip("Test skipped, requires PyQt5.")
mne/conftest.py:    from PyQt5.QtWidgets import QApplication, QMainWindow
mne/gui/__init__.py:    from PyQt5.QtWidgets import QApplication
mne/gui/__init__.py:        from PyQt5 import QtGui
mne/gui/_ieeg_locate_gui.py:from PyQt5 import QtCore, QtGui, Qt
mne/gui/_ieeg_locate_gui.py:from PyQt5.QtCore import pyqtSlot
mne/gui/_ieeg_locate_gui.py:from PyQt5.QtWidgets import (QMainWindow, QGridLayout,
mne/utils/check.py:        from PyQt5.Qt import PYQT_VERSION_STR as version
mne/utils/check.py:        warn('macOS users should use PyQt5 >= 5.10 for GUIs, got %s. '
mne/utils/check.py:             '    pip install "PyQt5>=5.10,<5.14"\n'
mne/utils/config.py:        PyQt5:         5.15.0
mne/utils/config.py:                     'PyQt5', 'ipympl', 'pooch', '', 'mne_bids', 'mne_nirs',
mne/utils/config.py:            elif mod_name == 'PyQt5':
mne/viz/_scraper.py:        from PyQt5.QtWidgets import QApplication
mne/viz/backends/tests/_utils.py:    """Check if PyQt5 is installed."""
mne/viz/backends/tests/_utils.py:        import PyQt5  # noqa: F401
mne/viz/raw.py:    fig : matplotlib.figure.Figure | ``PyQt5.QtWidgets.QMainWindow``
mne/viz/utils.py:    """Protect against PyQt5 exiting on event-handling errors."""
mne/viz/utils.py:        from PyQt5.QtWidgets import QApplication
tools/azure_dependencies.sh:    python -m pip install --progress-bar off --upgrade --pre --only-binary ":all:" --no-deps --extra-index-url https://www.riverbankcomputing.com/pypi/simple PyQt5 PyQt5-sip PyQt5-Qt5
tools/circleci_dependencies.sh:echo "Working around PyQt5 bugs"
tools/github_actions_dependencies.sh:   echo "PyQt5"
tools/github_actions_dependencies.sh:   pip install $STD_ARGS --pre --only-binary ":all:" --no-deps --extra-index-url https://www.riverbankcomputing.com/pypi/simple PyQt5 PyQt5-sip PyQt5-Qt5
```

</details>

**Comment by larsoner** [comment]
As a first step let's make the only change CIs just to add `qtpy` to the CI requirements, keeping PyQt5 as the lib that actually gets installed. Then this PR's purpose can be to remove `PyQt5` uses in our code in favor of some `pyqt` abstraction. Once that's green we can merge.

Then the next PR can be to improve CIs by adding PySide2 / PySide 6 / PyQt6 runs -- this will require a bit of work and thinking as we probably *only* want to test things that use our new abstraction (Brain, iEEG GUI, etc.). The easiest way to do this is maybe by making a `pytest.mark.qt` or so to signify that it's a Qt test, then we can `pytest -m qt` in these (hopefully quick) framework-only CI runs. Or maybe we get this for free by using something like `pytest-qt` with a `qtbot` fixture, in which case we should just add it to the appropriate set of tests and use it.

**Comment by GuillaumeFavelier** [comment]
Also I think there is no hurry. I jumped on this because I am interested but I can wait for the release.

**Comment by larsoner** [comment]
Agreed we should merge this right after release. So actually if the qtpy abstraction is quick and some commit is green before release, it's okay to roll the CI changes into this PR, too. It will be obvious what CIs and infrastructure have been added to make it work I think.

**Comment by cbrnr** [comment]
At the time I made the switch, QtPy did not support Qt6, around a year after Qt6 had been released. I mainly wanted to upgrade to Qt6 because I found a bug in Qt5's font rendering (which looked really bad).

Moreover, PySide6 is the official Python package from Qt. Besides some nice features (which I currently don't use, but snake case names is tempting), it is licensed under the LGPL as opposed to PyQt6's GPL. I don't think that this prohibits PyQt6 from using together with BSD/MIT-style licenses, but LGPL is definitely less restrictive. PySide6 is available for all three platforms as binary wheels on PyPI. For these reasons, I didn't see the necessity to (1) support PyQt6 and (2) introduce yet another layer of abstraction (QtPy).

So far, I'm quite happy with the transition. To be honest, not that much changed since Qt seems to be in maintenance mode anyway (they seem to focus on QtQuick instead), but at least they still support PySide6 officially and fix bugs.

**Comment by cbrnr** [comment]
Also, regarding loading icons, I load SVGs directly (using a theme) in MNELAB, e.g. https://github.com/cbrnr/mnelab/blob/main/mnelab/mainwindow.py#L80

I'd be happy to discuss all Qt-related things with you – we'd probably both benefit a lot from each other because each of us has already solved different problems with Qt.

**Comment by GuillaumeFavelier** [comment]
Thanks for chiming in @hoechenberger, @cbrnr, your feedback is appreciated :)

Personally, I am interested in `qtpy` because I want to offer the flexibility to choose whatever Qt api you want/like/use but maybe not the majority of `mne-python` users or devs care about this matter at all. After all, `mne-python` sticks to `PyQt5` for a while now but I am not sure if it's really by choice considering how sometimes it's difficult to migrate to another API. At least, this PR makes it easier to switch, especially in the long run.

> So far, I'm quite happy with the transition.

From another perspective, maybe the transition was not painful because you were using `qtpy` in the first place? :thinking: 

> PySide6 is the official Python package from Qt
> PySide6 is available for all three platforms as binary wheels on PyPI

Those are all good points, that's why I am not opposed to choose `PySide6` by default for example.

> yet another layer of abstraction (QtPy)

In my opinion, this is a small layer and very comfortable to use as a dev :sweat_smile: 

This is just my opinion but of course there are also considerations about licensing and maintenance cost for the project. I'll be happy to help no matter what we decide.

**Comment by GuillaumeFavelier** [comment]
> I load SVGs directly (using a theme) 

I'll take a look right away! :star_struck:

**Comment by cbrnr** [comment]
> Personally, I am interested in `qtpy` because I want to offer the flexibility to choose whatever Qt api you want/like/use but maybe not the majority of `mne-python` users or devs care about this matter at all. After all, `mne-python` sticks to `PyQt5` for a while now but I am not sure if it's really by choice considering how sometimes it's difficult to migrate to another API. At least, this PR makes it easier to switch, especially in the long run.

I agree, I think almost no one will care whether the GUI uses PySide6 or PyQt6. Having said that, there is not really any reason why one would choose PyQt6 over PySide6, so we might as well decide to stick with PySide6 in the future. All the other choices offered via QtPy are older versions of Qt, and I'm not sure why we should keep supporting those.

> From another perspective, maybe the transition was not painful because you were using `qtpy` in the first place? 🤔

Maybe. It's basically just replacing `qtpy` with `PySide6` in the imports.

> Those are all good points, that's why I am not opposed to choose `PySide6` by default for example.

I think this would indeed be a sensible choice.

> In my opinion, this is a small layer and very comfortable to use as a dev 😅

It is, yes, but if there is no real benefit then I'd rather not use it, no matter how small it is. The more you add, the more potential issues you will encounter at some point.

File: mne/utils/check.py
Comment (hoechenberger) [inline mne/utils/check.py:800]: Could you use f-strings here please? I think we do that in most of our new code these days.

Comment (cbrnr) [reply mne/utils/check.py:800]: Do you mean using f-strings for implicit concatenation (e.g. `f'Please upgrade e.g. with:\n\n'` or f-strings to replace the `%` operator? I think the latter should definitely be replaced, but I'm not sure the first is necessary (but it reads nicer).

Comment (hoechenberger) [reply mne/utils/check.py:800]: I mean the latter, right now I see an unholy mix of f-strings and `%s`

Comment (GuillaumeFavelier) [reply mne/utils/check.py]: I suggest [d148377](https://github.com/mne-tools/mne-python/pull/10430/commits/d1483778807b9cd15e208e360b6deed14620bb54)

Comment (cbrnr) [reply mne/utils/check.py]: Very nice, this avoids the awkward f-string which doesn't contain `{}`!

Comment (hoechenberger) [reply mne/utils/check.py]: thx!

**Comment by cbrnr** [comment]
Good idea to discuss the resource system in a separate PR. I'd switch to PySide6 directly (i.e. without `QtPy`) – but this means that there needs to be a `pyside6` package in conda-forge before the next release, right?

**Comment by cbrnr** [comment]
Maybe so. For reference: https://github.com/conda-forge/qt-main-feedstock/pull/14

**Comment by larsoner** [comment]
I don't think we need to wait for conda forge to have a package to prefer it internally (as long as we can nicely fall back to PyQt6/5). People who use pip will still get the benefits, and we're ready whenever conda comes around. So I say keep going @GuillaumeFavelier !

**Comment by GuillaumeFavelier** [comment]
I can continue working on this and switch from `PyQt5` to `PySide6` of course. But is that okay for `environment.yml` and `mne-installers` ? I might have to rely on `pip` if the package is not available on `conda-forge`.

> as long as we can nicely fall back to PyQt6/5

I'm not sure to understand what you mean

**Comment by larsoner** [comment]
Sorry, I missed the arguments against `qtpy` above. I think they are mostly not relevant for us and that we should transition to `qtpy` rather than plain PySide6. In addition to allowing PyQt5/PyQt6/PySide2/PySide6 already now (right?), if there is ever a PySide7/QtPy7, it will probably reduce the transition pain. To me it's the easiest, most compatible option. The only real downside is that we can't make a quick transition to PySide7 or something later, but we (MNE) never go to the latest-and-greatest right away -- our `requirements` tend to support things to 2 years back as much as possible.

**Comment by hoechenberger** [comment]
> but we (MNE) never go to the latest-and-greatest right away

*sad noises 😭*

**Comment by larsoner** [comment]
> provided that there's a conda-forge package of PySide6, there is no reason to support PyQt5/PyQt6/PySide2 anymore.

I disagree, we don't want to force people to update, even if an update is available on all platforms.

**Comment by cbrnr** [comment]
Yes, I don't see the point in maintaining old versions. Qt6 has been released 15 months ago and we're still OK with people using the old Qt5? Issues will only be fixed in Qt6, e.g. the font rendering bug is here to stay in Qt5. So once the conda-forge package is out I really don't understand why we don't want to force people to upgrade. For me, it is already annoying to support anything else other than PySide6.

But OK, if you disagree I'm also fine with adding QtPy - but keep in mind this is yet another package we're adding to our dependency pile.

**Comment by larsoner** [comment]
> Qt6 has been released 15 months ago and we're still OK with people using the old Qt5?

Yes, upgrades often cause problems. Even for me as an experienced dev, I tried testing `mnelab` recently and it required PySide6, but installing this broke PyVistaQt. I couldn't have both installed at once. Things like this are not uncommon.

We have a standard policy of supporting libraries 2 years out (as much as possible), and in this case, Qt6 hasn't been available directly via `conda` (our preferred installation method) *at all* yet, so even though PySide6 has been out 15 months, I would vote for considering the 2-year clock to have not started yet on this library (!).

> For me, it is already annoying to support anything else other than PySide6.

I think if we go the qtpy route, it will minimize our pain. And I think we need to do it for our users.

> keep in mind this is yet another package we're adding to our dependency pile.

Agreed it's sad to add another dependency, but here I think it's a necessary evil.

Let's move on with qtpy.

**Comment by GuillaumeFavelier** [comment]
Thank you everyone for taking the time to interact in here. All your suggestions are valuable so I'll do my best to converge to something satifisfying :+1: 

---

Considering the latest comments, this is what I suggest:
* I will update this PR to 'Add support to QtPy'
* Then I will open an issue about the choice of the default Qt API

At the moment, I am busy working on Qt resource management (icon & them) but once this is taken care of, I'll go back on this.

Sorry for the noise, I think I miscalculated how ambitious this PR would be :pray:

**Comment by hoechenberger** [comment]
Hello, I just wanted to chime in again and say that I've discussed this issue with @GuillaumeFavelier personally last weekend, and I now also agree that we should probably go with QtPy. It seems to be the most sustainable option in terms of future maintenance work.

**Comment by GuillaumeFavelier** [comment]
Locally, I was able to try `coreg` and `plot_stc` on c2b4f7eb9016d382a5e1e43af9a554c9cb1c1a7b with `PySide6` and `PyQt6` and it works as I would expect. I did not find any bug for now.

When the window closes I do see the following in the log:

```
qt.qpa.xcb: QXcbConnection: XCB error: 3 (BadWindow), sequence: 5660, resource id: 15444581, major code: 40 (TranslateCoords), minor code: 0
```

**Review by larsoner — CHANGES_REQUESTED** [review]

File: requirements.txt
Comment (GuillaumeFavelier) [reply requirements.txt:11]: I reverted c2b4f7eb9016d382a5e1e43af9a554c9cb1c1a7b right away because I don't think we want to change that now

**Comment by larsoner** [comment]
> I reverted c2b4f7e right away because I don't think we want to change [the Qt version pinning] now

Sure, seems reasonable to move to `qtpy` first but keep using PyQt5 everywhere as far as our `requirements.txt` and `environment.yml` go.

In this PR or a follow-up one, we could add a CI job that runs GUI + viz tests (only, for speed) with a matrix of PyQt5, PyQt6, PySide2, and PySide6 (probably all just on linux, again for speed?). WDYT?

Then separately we can look to see if we can fix HiDPI on macOS. Once that works, we can deal with `environment.yml` and `requirements.txt`. This can all land in 1.1 so that we have months to sort it out.

**Comment by larsoner** [comment]
> I think this goes in pair with environment.yml and requirements.txt because it's very messy to test with multiple APIs installed or we need to hack something in the job itself to ensure there is only one.

There are already places we give complete, separate specs of what to install, like https://github.com/mne-tools/mne-python/blob/main/tools/azure_dependencies.sh and the GH actions equivalent. I would "just" add a new `else if` clause in those with some `QT_API` arg. Then we define the `QT_API` arg in some tests.

Should I give it a shot in this PR? I'm thinking Azure probably has the lightest load currently... actually I'd probably try reenabling [appveyor](https://www.appveyor.com/docs/lang/python/) since I think it does still work but we don't use it. We could even test Ubuntu and Windows if we wanted, but I'd probably just do half on one and half on the other for now.

**Comment by GuillaumeFavelier** [comment]
> Should I give it a shot in this PR?

If you want, I just pushed so I'm okay with it.

**Comment by larsoner** [comment]
@GuillaumeFavelier should we merge https://github.com/mne-tools/mne-python/pull/10495 once it's ready, then rebase/merge-with-main in this PR?

**Review by larsoner — APPROVED** [review]
@GuillaumeFavelier feel free to mark as ready for review and merge if you're happy!

It looks like PyQtGraph might already support the four Qt bindings here (TBC) but assuming that's the case, I *think* all we need is for mne-qt-browser to use the same `qtpy` abstraction framework, and then we'll have the whole MNE plotting stack (matplotlib, pyvistaqt, mne-qt-browser/pyqtgraph) compatible with all four frameworks. Would you be up for making a PR to mne-qt-browser with the same abstraction changes? Bonus points if you want to try to make a similar Azure pipelines update. (I think it's okay just to run it in the `mne-dev` case over there, which would then pick up the changes from this PR.)

**Comment by GuillaumeFavelier** [comment]
The `Test Qt` job skips some tests while testing `PyQt6` and `PySide2`:

```
mne/viz/_brain/tests/test_brain.py::test_layered_mesh[pyvistaqt] SKIPPED (Test skipped, requires pyvistaqt) [  2%]
mne/viz/_brain/tests/test_brain.py::test_single_hemi[pyvistaqt-lh] SKIPPED (Test skipped, requires pyvistaqt) [  5%]
mne/viz/_brain/tests/test_brain.py::test_single_hemi[pyvistaqt-rh] SKIPPED (Test skipped, requires pyvistaqt) [  8%]
mne/viz/_brain/tests/test_brain.py::test_brain_screenshot[pyvistaqt] SKIPPED (Test skipped, requires pyvistaqt) [ 11%]
mne/viz/_brain/tests/test_brain.py::test_foci_mapping[pyvistaqt] SKIPPED (Test skipped, requires pyvistaqt) [ 14%]
mne/viz/_brain/tests/test_brain.py::test_brain_gc[pyvistaqt] SKIPPED (Test skipped, requires pyvistaqt) [ 17%]
mne/viz/_brain/tests/test_brain.py::test_brain_routines[pyvistaqt] SKIPPED (Test skipped, requires pyvistaqt) [ 20%]
...
```


And the  `PySide2` section of `mne sys_info` does not find `pyvistaqt`:

<details>

```
Platform:            Linux-5.13.0-1017-azure-x86_64-with-glibc2.31
Python:              3.10.4 (main, Mar 24 2022, 14:07:25) [GCC 9.4.0]
Executable:          /opt/hostedtoolcache/Python/3.10.4/x64/bin/python
CPU:                 x86_64: 2 cores
Memory:              6.8 GB

mne:                 1.1.dev0
                     •/home/vsts/work/1/s/mne
numpy:               1.21.5 {blas=openblas, lapack=openblas}
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/numpy
scipy:               1.8.0
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/scipy
matplotlib:          3.5.1 {backend=QtAgg}
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/matplotlib

sklearn:             1.0.2
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/sklearn
numba:               0.55.1
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/numba
nibabel:             3.2.2
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/nibabel
nilearn:             0.9.0
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/nilearn
dipy:                1.5.0
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/dipy
cupy:                Not found
pandas:              1.4.2
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/pandas
pyvista:             0.34.0 {OpenGL 4.5 (Core Profile) Mesa 21.2.6 via llvmpipe (LLVM 12.0.0, 256 bits)}
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/pyvista
pyvistaqt:           Not found
ipyvtklink:          0.2.2
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/ipyvtklink
vtk:                 9.1.0.rc1
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/vtkmodules
qtpy:                2.0.1 {PySide2=5.15.2.1}
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/qtpy
ipympl:              Not found
pooch:               v1.6.0
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/pooch

mne_bids:            Not found
mne_nirs:            Not found
mne_features:        Not found
mne_qt_browser:      0.2.6
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/mne_qt_browser
mne_connectivity:    Not found

sphinx:              4.5.0
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/sphinx
sphinx_gallery:      0.10.1
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/sphinx_gallery
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/pytest
nbclient:            0.5.13
                     •/opt/hostedtoolcache/Python/3.10.4/x64/lib/python3.10/site-packages/nbclient
qtpy:                2.0.1 {PySide2=5.15.2.1}
```

</details>

**Comment by larsoner** [comment]
Good catch. Can you replicate locally?

**Comment by larsoner** [comment]
Maybe we need the dev version of pyvistaqt / maybe it needs a new release with some of your fixes?

**Comment by GuillaumeFavelier** [comment]
> Can you replicate locally?

I'm on it!

**Comment by GuillaumeFavelier** [comment]
I narrowed the issue down to:

```
# Check whether a specific PyQt implementation was chosen
try:
    import vtkmodules.qt
    PyQtImpl = vtkmodules.qt.PyQtImpl
except ImportError:
    pass
```

Somehow, `vtkmodules.qt.PyQtImpl` is still set to `PyQt5` forcing `pyvistaqt/rwi.py` to load `PyQt5` causing the error preventing `pyvistaqt` to be imported.

**Comment by GuillaumeFavelier** [comment]
And all of this happens because `PyQt6` is not officially supported in `vtk` yet...

I modified the following locally and it works:

```diff
# Has an implementation has been imported yet?
-for impl in ["PySide6", "PyQt5", "PySide2", "PyQt4", "PySide"]:
+for impl in ["PyQt6", "PySide6", "PyQt5", "PySide2", "PyQt4", "PySide"]:
    if impl in sys.modules:
        PyQtImpl = impl
        break
```

https://gitlab.kitware.com/vtk/vtk/-/blob/master/Wrapping/Python/vtkmodules/qt/__init__.py#L27-31

**Comment by GuillaumeFavelier** [comment]
I guess I have to open this PR rather sooner than later :sweat_smile: 

Meanwhile, do you think it's worth patching in `pyvistaqt` @larsoner ?

**Comment by larsoner** [comment]
IIRC VTK 9.2 will already support PyQt6. I think it's supposed to come out in a month. So perhaps for now we should just omit the PyQt6 support / CI entry.

An option in the interim would be to get an updated pyvista wheel with 9.2dev that supports PyQt6. But it doesn't seem worth it.

So that covers PyQt6. Why isn't PySide2 working though?

**Comment by GuillaumeFavelier** [comment]
> IIRC VTK 9.2 will already support PyQt6

This is good to know, thank you for the info!

> So that covers PyQt6. Why isn't PySide2 working though?

I will update the PR to disable PyQt6 testing then. I'll continue to investigate

**Comment by larsoner** [comment]
> I will update the PR to disable PyQt6 testing then.

On second thought, let's just leave the PyQt6 CI entry in here. It at least tests matplotlib, and eventually will get mne-qt-browser, then finally VTK. And the skips work in the meantime.

So I think all that is left to do is figure out PySide2...

**Comment by GuillaumeFavelier** [comment]
For `PySide2`, the error is slightly different:

```
❯ python
Python 3.9.12 (main, Apr  5 2022, 06:56:58) 
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvistaqt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/guillaume/source/anaconda3/envs/test/lib/python3.9/site-packages/pyvistaqt/__init__.py", line 31, in <module>
    from .plotting import BackgroundPlotter, MainWindow, MultiPlotter, QtInteractor
  File "/home/guillaume/source/anaconda3/envs/test/lib/python3.9/site-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/home/guillaume/source/anaconda3/envs/test/lib/python3.9/site-packages/pyvistaqt/plotting.py", line 76, in <module>
    from .rwi import QVTKRenderWindowInteractor
  File "/home/guillaume/source/anaconda3/envs/test/lib/python3.9/site-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/home/guillaume/source/anaconda3/envs/test/lib/python3.9/site-packages/pyvistaqt/rwi.py", line 143, in <module>
    from PyQt5.QtWidgets import QWidget
  File "/home/guillaume/source/anaconda3/envs/test/lib/python3.9/site-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
ModuleNotFoundError: No module named 'PyQt5.QtWidgets'
>>> from qtpy import API
>>> API
'pyside2'
>>> 
```

This time, `PySide2` is in the list of `vtkmodules.qt` but `PyQt5` is checked first and for an unknown reason:

```
import sys

# PyQtImpl can be set by the user
PyQtImpl = None

# Has an implementation has been imported yet?
for impl in ["PySide6", "PyQt5", "PySide2", "PyQt4", "PySide"]:
    if impl in sys.modules:
        PyQtImpl = impl
        break
```

`PyQtImpl` is set to `PyQt5` even though I uninstalled it :thinking:

**Comment by larsoner** [comment]
> PyQtImpl is set to PyQt5 even though I uninstalled it 

If you're on linux, perhaps you have one in the system `dist`, like `python3-pyqt5` or whatever?

**Comment by GuillaumeFavelier** [comment]
I investigated and it is caused by `qtpy`. Setting `QT_API` should fix it.

**Comment by larsoner** [comment]
> Setting QT_API should fix it.

Can you open an upstream bug report with them? In the meantime, let's set this env var in just the PySide2 calls, then. (I can push a commit for this if you don't get what I mean.) And let's reenable PyQt6 -- sorry for the suggestion to disable earlier! Give me 1h to try to get a working VTK 9.2dev build to see if I can get PyQt6 fully working, but even if we don't, we can still test the non-3D stuff with that run...

**Comment by GuillaumeFavelier** [comment]
> Can you open an upstream bug report with them?

Sorry for the noise. Technically, `qtpy` is right because it tries to import `PyQt5` and it properly detects `PySide2` afterwards. *That's when `PyQt5` ends up in `sys.modules`* and then is detected by `vtk` (because of the order in `sysmodules.qt`).

The real problem is that it **should not** be possible to `import PyQt5` after uninstalling it right? Locally, it's still there for me:

```
python3.9/site-packages/PyQt5 
❯ tree   
.
├── bindings
└── Qt5
    ├── plugins
    ├── qml
    │   ├── Qt
    │   │   ├── labs
    │   │   └── test
    │   └── QtQuick
    └── qsci
        └── api

10 directories, 0 files
```

When I delete this directory manually, the issue is resolved because `PyQt5` is not present in `sys.modules`. Setting `QT_API` prevents `qtpy` from importing `PyQt5`, that's why it also works for me.

I can also guess that it works with `PySide6` because it has higher priority than `PyQt5` in `vtkmodules.qt`.

At this point, I don't know if it's only me or this is something related to `conda` or `pip` :sweat_smile:

**Comment by larsoner** [comment]
> At this point, I don't know if it's only me or this is something related to conda or pip

Does `pip list | grep -i pyqt5` yield anything? Maybe there is some other package like https://pypi.org/project/pyqt5-tools/ is installed?

**Comment by GuillaumeFavelier** [comment]
One moment, I will double-check because I modified the env too much :+1:

**Comment by larsoner** [comment]
... given that this can happen, though, it might still make sense to make a PR to `qtpy` to go beyond just a `in sys.modules` check, and actually try to import something that should be there, like `QtCore`, and only accept it as "imported" if it works.

**Comment by GuillaumeFavelier** [comment]
The check of `sys.modules` is done in `vtk`, `qtpy` is correct and tries to import `QtCore`, that's why it detects `PySide2` in the end:

https://github.com/mne-tools/mne-python/pull/10430#issuecomment-1090238118

**Comment by larsoner** [comment]
> https://gitlab.kitware.com/vtk/vtk/-/blob/master/Wrapping/Python/vtkmodules/qt/__init__.py#L27-31

Argh, I missed that this was already `master`! It would be good to open a PR to VTK to just add `PyQt6` now that 9.2 will support it. They hopefully/maybe just forgot to update this Python code along the way. It's also possible that more changes will be required beyond just adding it, but at least we can get the ball rolling with them by making a PR to add PyQt6 to the list. Feel free to ping me over there (still as @larsoner )

**Comment by larsoner** [comment]
> qtpy is correct and tries to import QtCore,

I would also port this over to VTK and see what they say. In a worst case we have to remove it and just live with the `in sys.modules` check, but I assume qtpy did it for a good reason (maybe the one we're hitting here actually...)

**Comment by larsoner** [comment]
... FYI here is a little diff I'm going to apply to VTK `master` and do a `manylinux` docker wheel build of 9.2dev / master with the following patch applied:
```diff
diff --git a/Wrapping/Python/vtkmodules/qt/__init__.py b/Wrapping/Python/vtkmodules/qt/__init__.py
index 2b5b2f73fb..cb7c27a308 100644
--- a/Wrapping/Python/vtkmodules/qt/__init__.py
+++ b/Wrapping/Python/vtkmodules/qt/__init__.py
@@ -19,16 +19,24 @@ For more information, see QVTKRenderWidgetConeExample() in the file
 QVTKRenderWindowInteractor.py.
 """
 
+import importlib
 import sys
 
 # PyQtImpl can be set by the user
 PyQtImpl = None
 
 # Has an implementation has been imported yet?
-for impl in ["PySide6", "PyQt5", "PySide2", "PyQt4", "PySide"]:
+for impl in ["PySide6", "PyQt6", "PyQt5", "PySide2", "PyQt4", "PySide"]:
     if impl in sys.modules:
-        PyQtImpl = impl
-        break
+        # Sometimes an attempted import can be crufty (e.g., unclean
+        # uninstalls of PyQt5), so let's try to import the actual functionality
+        try:
+            importlib.import_module(impl + '.QtCore')
+        except Exception:
+            pass
+        else:
+            PyQtImpl = impl
+            break
 
 # QVTKRWIBase, base class for QVTKRenderWindowInteractor,
 # can be altered by the user to "QGLWidget" or "QOpenGLWidget"
```
It's probably going to take a couple of hours to actually build, then I'll upload the wheel to OSF, and we can try it here!

**Comment by GuillaumeFavelier** [comment]
> It's probably going to take a couple of hours to actually build, then I'll upload the wheel to OSF, and we can try it here!

Thank you for the help! I think the updated test is more robust indeed. 

> Does pip list | grep -i pyqt5 yield anything?

Sorry for the delay, I purged my `pip` cache:

```
mne-python on  mnt/qtpy via 🐍 v3.9.12 via 🅒 test 
❯ pip list | grep -i qt   
mne-qt-browser                0.2.6
pyqtgraph                     0.12.4
pyvistaqt                     0.8.0
qtconsole                     5.3.0
QtPy                          2.0.1
sphinxcontrib-qthelp          1.0.3
```
```
mne-python on  mnt/qtpy via 🐍 v3.9.12 via 🅒 test 
❯ pip list | grep -i pyside2
PySide2                       5.15.2.1
```

**Comment by GuillaumeFavelier** [comment]
At the moment, there are some timeouts for:

```
mne-python on  mnt/qtpy via 🐍 v3.9.7 via 🅒 pyside2 took 18s 
❯ pip install -i "https://pypi.anaconda.org/scipy-wheels-nightly/simple" pandas            
Looking in indexes: https://pypi.anaconda.org/scipy-wheels-nightly/simple
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.anaconda.org', port=443): Read timed out. (read timeout=15)")': /scipy-wheels-nightly/simple/pandas/
```

I will try again later.

**Comment by larsoner** [comment]
Thanks @GuillaumeFavelier !

**Pull request merged**
